buffer/src/main/java/io/netty/buffer/api/CompositeBuffer.java

```diff
@@ -0,0 +1,2028 @@
+        return new CompositeBuffer(allocator, EMPTY_BUFFER_ARRAY, COMPOSITE_DROP);
+    }
+
+    /**
+     * Check if the given buffer is a {@linkplain #compose(BufferAllocator, Send...) composite} buffer or not.
+     * @param composite The buffer to check.
+     * @return {@code true} if the given buffer was created with {@link #compose(BufferAllocator, Send...)},
+     * {@code false} otherwise.
+     */
+    public static boolean isComposite(Buffer composite) {
+        return composite.getClass() == CompositeBuffer.class;
+    }
+
+    private static Buffer[] filterExternalBufs(Stream<Buffer> refs) {
+        // We filter out all zero-capacity buffers because they wouldn't contribute to the composite buffer anyway,
+        // and also, by ensuring that all constituent buffers contribute to the size of the composite buffer,
+        // we make sure that the number of composite buffers will never become greater than the number of bytes in
+        // the composite buffer.
+        // This restriction guarantees that methods like countComponents, forEachReadable and forEachWritable,
+        // will never overflow their component counts.
+        // Allocating a new array unconditionally also prevents external modification of the array.
+        Buffer[] bufs = refs
+                .filter(CompositeBuffer::discardEmpty)
+                .flatMap(CompositeBuffer::flattenBuffer)
+                .toArray(Buffer[]::new);
+        // Make sure there are no duplicates among the buffers.
+        Set<Buffer> duplicatesCheck = Collections.newSetFromMap(new IdentityHashMap<>());
+        duplicatesCheck.addAll(Arrays.asList(bufs));
+        if (duplicatesCheck.size() < bufs.length) {
+            IllegalArgumentException iae = new IllegalArgumentException(
+                    "Cannot create composite buffer with duplicate constituent buffer components.");
+            for (Buffer buf : bufs) {
+                try {
+                    buf.close();
+                } catch (Exception closeExc) {
+                    iae.addSuppressed(closeExc);
+                }
+            }
+            throw iae;
+        }
+        return bufs;
+    }
+
+    private static boolean discardEmpty(Buffer buf) {
+        if (buf.capacity() > 0) {
+            return true;
+        } else {
+            // If we filter a buffer out, then we must make sure to close it since it's ownership was sent to us.
+            buf.close();
+            return false;
+        }
+    }
+
+    private static Stream<Buffer> flattenBuffer(Buffer buf) {
+        if (buf instanceof CompositeBuffer) {
+            // Extract components so composite buffers always have non-composite components.
+            var composite = (CompositeBuffer) buf;
+            return Stream.of(composite.bufs);
+        }
+        return Stream.of(buf);
+    }
+
+    private CompositeBuffer(BufferAllocator allocator, Buffer[] bufs, Drop<CompositeBuffer> drop) {
+        super(drop);
```

Comment:
I think there is a risk here.    The super call has to happen before any other call, which means it cannot be part of the try catch block.   If the super call throws, the bufs will be leaked.   

I think the callers of this ctor need to either have externally do the try-catch, or some how the bufs get initialized after the ctor finishes.

Reply:
I can only think of `StackOverflowError` and `OutOfMemoryError` as possibilities for `super` to throw. If those occur, we are pretty much screwed regardless.